qermit/coherent_pauli_checks/monochromatic_convex_subdag.py

```diff
@@ -0,0 +1,147 @@
+from itertools import permutations
+from typing import Any
+
+import networkx as nx  # type: ignore
+
+
+def _subdag_nodes(subdag: int, node_subdag: dict[Any, int]) -> list[Any]:
+    """Get all nodes in a given sub-DAG.
+
+    :param subdag: The sub-DAG whose nodes should be retrieved.
+    :param node_subdag: Map from node to the sub-DAG to which the
+        node belongs.
+    :return: List of nodes belonging to given sub-DAG.
+    """
+    return [node for node, s in node_subdag.items() if s == subdag]
+
+
+def _subdag_predecessors(
+    dag: nx.DiGraph, subdag: int, node_subdag: dict[Any, int]
+) -> list[Any]:
+    """Retrieve all nodes not in given sub-DAG with successors in sub-DAG.
+
+    :param dag: Directed Acyclic Graph.
+    :param subdag: Sub-DAG to retrieve predecessors of.
+    :param node_subdag: Map from node to the sub-DAG to which it belongs.
+    :return: Nodes with successors in given sub-DAG.
+    """
+    subdag_nodes = _subdag_nodes(subdag=subdag, node_subdag=node_subdag)
+    return sum(
+        [
+            [
+                predecessor
+                for predecessor in dag.predecessors(node)
+                # Exclude nodes in subdag.
+                if predecessor not in subdag_nodes
+            ]
+            for node in subdag_nodes
+        ],
+        start=[],
+    )
+
+
+def _subdag_successors(
+    dag: nx.DiGraph, subdag: int, node_subdag: dict[Any, int]
```

Comment:
All these functions take `node_subdag` as an argument but if implemented as a class this wouldn't be necessary. Just a suggestion.

Reply:
Yeah, I went back and forth on this. I personally don't link classes which are effectively just on function. That's not quite the case here, but any class would in effect just be the function `get_monochromatic_convex_subdag`. I could probably be persuaded either way, but this doesn't feel like a class to me.